### PR TITLE
adapter: Provide linearizable boot timestamp

### PR DESCRIPTION
**Pull request by jkosh44**
Previously, we would provide a non-linearizable boot timestamp to `Catalog::open`. A non-linearizable timestamp is not written down, and therefore it's possible that a later environment boots up with a lower timestamp. We were unable to provide a linearizable timestamp due to a bootstrapping problem. We needed to look in the catalog to see what timestamp oracle implementation to use, but we needed to know which timestamp oracle implementation to use to get a linearizable timestamp to open the catalog.

Now that there's only a single timestamp oracle implementation, this bootstrapping problem has gone away. This commit updates the bootstrap process so that a linearizable timestamp is passed to `Catalog::open`. This is useful for things like adding audit log events in a migration, which require linearizable timestamps.

### Motivation
This PR adds a feature that has not yet been specified.

### Checklist

- [X] This PR has adequate test coverage / QA involvement has been duly considered.
- [X] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [X] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [X] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [X] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - There are no user-facing behavior changes.

### CODE REVIEW

**Comment by jkosh44** [comment]
@aljoscha do you agree with my commit message? Am I remembering things correctly?

File: src/adapter/src/coord.rs [hunk 1]
```diff
@@ -2943,39 +2938,35 @@ pub fn serve(
...
             get_initial_oracle_timestamps(&pg_timestamp_oracle_config).await?;
 
-        // A candidate for the boot_ts. Catalog::open will further advance this,
-        // based on the "now" timestamp, if/when needed.
-        let previous_ts = initial_timestamps
-            .get(&Timeline::EpochMilliseconds)
-            .cloned()
-            .unwrap_or_else(mz_repr::Timestamp::minimum);
-
+        // Insert an entry for the `EpochMilliseconds` timeline if one doesn't,
```
Comment (mjibson) [inline src/adapter/src/coord.rs:2941]:
```suggestion
        // Insert an entry for the `EpochMilliseconds` timeline if one doesn't exist,
```

**Review by mjibson — APPROVED** [review]

File: src/adapter/src/coord/timeline.rs [hunk 2]
```diff
@@ -193,15 +191,11 @@ impl Coordinator {
...
-        timestamp_oracle_impl: TimestampOracleImpl,
         pg_oracle_config: Option<PostgresTimestampOracleConfig>,
         global_timelines: &'a mut BTreeMap<Timeline, TimelineState<Timestamp>>,
     ) -> &'a mut TimelineState<Timestamp> {
         if !global_timelines.contains_key(timeline) {
-            info!(
-                "opening a new {:?} TimestampOracle for timeline {:?}",
-                timestamp_oracle_impl, timeline,
-            );
+            info!("opening a new TimestampOracle for timeline {:?}", timeline,);
```
Comment (aljoscha) [inline src/adapter/src/coord/timeline.rs:198]: maybe leave something like "opening new CRDB/postgres TimestampOracle ...", just for clarity?

**Review by aljoscha — APPROVED** [review]
Yes, I think the commit message is correct. Also, though, is there any place in the bootstrapping process that calls `apply_write`, maybe as part of writing/migrating entries of the audit log? Or we think that's not needed?

**Review by aljoscha — APPROVED** [review]
Yes, I think the commit message is correct. Also, though, is there any place in the bootstrapping process that calls `apply_write`, maybe as part of writing/migrating entries of the audit log? Or we think that's not needed?

File: src/adapter/src/coord/timeline.rs (hunk 2 above)
Comment (jkosh44) [reply src/adapter/src/coord/timeline.rs]: Good idea, done.

**Comment by jkosh44** [comment]
> Also, though, is there any place in the bootstrapping process that calls `apply_write`, maybe as part of writing/migrating entries of the audit log? Or we think that's not needed?

Technically yes, the bootstrap process will apply some builtin table updates which calls `apply_write`. It also creates tables which may call `apply_write`. Still I added an explicit call to `apply_write` with `boot_ts` once the catalog is fully opened.

**Pull request merged**